File: sample_repos/java/SED/3/src/main/java/ic/doc/templatemethod/Sequence.java
package ic.doc.templatemethod;

import java.util.Iterator;

public abstract class Sequence implements Iterable<Integer> {

    public final int term(int i) {
        if (i < 0) {
            throw new IllegalArgumentException("Not defined for indices < 0");
        }
        return generate_term(i);
    }


    public abstract int generate_term(int i);

    public Iterator<Integer> iterator() {
        return new SequenceIterator();
    }

    private class SequenceIterator implements Iterator<Integer> {

        private int index = 0;

        @Override
        public boolean hasNext() {
            return true;
        }

        @Override
        public Integer next() {
            return term(index++);
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException("Remove is not implemented");
        }
    }
}

File: sed_repos/python/SED3/src/template_method/sequence.py
from typing import Iterator


class Sequence:
    def term(self, i: int) -> int:
        if i < 0:
            raise ValueError("Not defined for indices < 0")
        return self.generate_term(i)

    def generate_term(self, i: int) -> int:
        raise NotImplementedError("generate_term method not implemented")

    def __iter__(self) -> Iterator[int]:
        return self.SequenceIterator(self)

    class SequenceIterator:
        def __init__(self, sequence):
            self.sequence = sequence
            self.index = 0

        def __iter__(self):
            return self

        def __next__(self) -> int:
            result = self.sequence.term(self.index)
            self.index += 1
            return result
